### PR TITLE
feat(Buttons): Color + Structure Redux [GM-237]

### PR DESCRIPTION
**Pull request by codecaaron**
### Buttons
 - Refactors all buttons to a single element (switching the focus states to `:before` elements - like anchors)
 - Removes `ButtonOutline` and `ButtonInner` in favor of a shared style object. 
 - Changes all colors to their new analogues (navy scale where gray was used).
 - Adds `danger` variants with new reds.
 - Updates active states as designs have drifted slightly.
 - Refactors implementation to be 100% variance and system props (with the ability to use padding safely!).
 - Messes around with using a `Selector.HOVER` enum for more complicated selector strings - to prevent drift and possibly guarantee ordering in the future.
 - Changes buttons to all display with inline-flex as most likely adjacent children (Icons) have system props and can be styled effectively without much customization.
 - Adds storybook examples of these new color variants.
 - Updates some aliases used in Inputs
 - Fixes downstream typings from the large refactor of buttons.

### Misc Changes
`templateVariants`
This helper wraps the variant creator function to make repeating color configurations much easier.  It accepts an array of any variant specific variables (in this case base and hover colors) and a template function that accepts those vars as a parameter and returns a valid `system.css` call.  

```tsx
// Before 
const myCoolVariant = system.variant({
  variants: {
    primary: {
      textColor: 'primary',
      '&:hover': {
        textColor: `primary-hover`,
      },
    },
    secondary: {
      textColor: 'secondary',
      '&:hover': {
        textColor: `secondary-hover`,
      },
    },
    danger: {
      textColor: 'danger',
      '&:hover': {
        textColor: `danger-hover`,
      },
    },
  },
});

// After 
const myCoolVariant = templateVariants(['primary', 'secondary', 'danger'], (color) => ({ 
   textColor: color,  
   '&:hover': { 
     textColor: `$[color}-hover`
    }
 })
);
```  

## Storybook 

Screen Shot 2021-06-05 at 2 27 38 PM
Screen Shot 2021-06-05 at 2 27 54 PM

## Figma 

Screen Shot 2021-06-05 at 1 03 55 PM

### PR Checklist

- [x] Related to designs: https://www.figma.com/file/LCIvvjJGa9OEIFNGlFrbUd/Refinement?node-id=123%3A1333 https://www.figma.com/file/N6TfoPI5OA7GanX6Zl8nWj/Buttons-%26-Links?node-id=729%3A44
- [x] Related to JIRA ticket: GM-237
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

### CODE REVIEW

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.0.1-alpha.eb8871.0<br/>@codecademy/gamut-illustrations@0.17.3-alpha.eb8871.0<br/>@codecademy/gamut-kit@0.4.9-alpha.eb8871.0<br/>@codecademy/gamut-labs@19.0.2-alpha.eb8871.0<br/>@codecademy/gamut-patterns@0.3.3-alpha.eb8871.0<br/>@codecademy/gamut-styles@11.0.1-alpha.eb8871.0<br/>@codecademy/gamut-tests@2.4.6-alpha.eb8871.0<br/>@codecademy/gamut@34.0.2-alpha.eb8871.0<br/>@codecademy/konamimojisplosion@0.2.37-alpha.eb8871.0<br/>@codecademy/markdown-overrides@0.5.49-alpha.eb8871.0<br/>@codecademy/styleguide@41.1.2-alpha.eb8871.0<br/>@codecademy/variance@0.12.2-alpha.eb8871.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60bba23f013623deab643a34--gamut-preview.netlify.app>https://60bba23f013623deab643a34--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60bba23f013623deab643a34>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.0.1-alpha.2fd033.0<br/>@codecademy/gamut-illustrations@0.17.3-alpha.2fd033.0<br/>@codecademy/gamut-kit@0.4.9-alpha.2fd033.0<br/>@codecademy/gamut-labs@19.0.2-alpha.2fd033.0<br/>@codecademy/gamut-patterns@0.3.3-alpha.2fd033.0<br/>@codecademy/gamut-styles@11.0.1-alpha.2fd033.0<br/>@codecademy/gamut-tests@2.4.6-alpha.2fd033.0<br/>@codecademy/gamut@34.0.2-alpha.2fd033.0<br/>@codecademy/konamimojisplosion@0.2.37-alpha.2fd033.0<br/>@codecademy/markdown-overrides@0.5.49-alpha.2fd033.0<br/>@codecademy/styleguide@41.1.2-alpha.2fd033.0<br/>@codecademy/variance@0.12.2-alpha.2fd033.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60bba9a35c3a87225170df69--gamut-preview.netlify.app>https://60bba9a35c3a87225170df69--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60bba9a35c3a87225170df69>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.0.1-alpha.f6df21.0<br/>@codecademy/gamut-illustrations@0.17.3-alpha.f6df21.0<br/>@codecademy/gamut-kit@0.4.9-alpha.f6df21.0<br/>@codecademy/gamut-labs@19.0.2-alpha.f6df21.0<br/>@codecademy/gamut-patterns@0.3.3-alpha.f6df21.0<br/>@codecademy/gamut-styles@11.0.1-alpha.f6df21.0<br/>@codecademy/gamut-tests@2.4.6-alpha.f6df21.0<br/>@codecademy/gamut@34.0.2-alpha.f6df21.0<br/>@codecademy/konamimojisplosion@0.2.37-alpha.f6df21.0<br/>@codecademy/markdown-overrides@0.5.49-alpha.f6df21.0<br/>@codecademy/styleguide@41.1.2-alpha.f6df21.0<br/>@codecademy/variance@0.12.2-alpha.f6df21.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60bbaf254618850f9eb2078b--gamut-preview.netlify.app>https://60bbaf254618850f9eb2078b--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60bbaf254618850f9eb2078b>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.20-alpha.467e4b.0<br/>@codecademy/gamut-labs@19.4.2-alpha.467e4b.0<br/>@codecademy/gamut@34.5.2-alpha.467e4b.0<br/>@codecademy/styleguide@41.7.2-alpha.467e4b.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c17c34c4a940857e1578b7--gamut-preview.netlify.app>https://60c17c34c4a940857e1578b7--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c17c34c4a940857e1578b7>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.20-alpha.6eeb5c.0<br/>@codecademy/gamut-labs@19.4.2-alpha.6eeb5c.0<br/>@codecademy/gamut@34.5.2-alpha.6eeb5c.0<br/>@codecademy/styleguide@41.7.2-alpha.6eeb5c.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c22d8088178b008aca6e50--gamut-preview.netlify.app>https://60c22d8088178b008aca6e50--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c22d8088178b008aca6e50>Deploy Logs</a></p>

File: packages/gamut/src/Button/shared.ts [hunk 1]
```diff
@@ -1,72 +1,106 @@
...
+  template: (colors: Variant) => Styles
+) => {
+  const variantConfig = {} as Record<Variant, ReturnType<typeof template>>;
+  variants.forEach((key: Variant) => {
+    variantConfig[key] = template(key);
+  });
+  return system.variant({
+    defaultVariant: variants[0],
+    variants: variantConfig,
+  });
```
Comment (dreamwasp) [inline packages/gamut/src/Button/shared.ts:38]: so .. cool... 🤩

**Review by dreamwasp — APPROVED** [review]
image
 dangggggggg

**Review by jakemhiller — APPROVED** [review]

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.24-alpha.2d6ac2.0<br/>@codecademy/gamut-labs@19.5.2-alpha.2d6ac2.0<br/>@codecademy/gamut@34.6.2-alpha.2d6ac2.0<br/>@codecademy/styleguide@41.7.8-alpha.2d6ac2.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c7f03762fef53ee280a924--gamut-preview.netlify.app>https://60c7f03762fef53ee280a924--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c7f03762fef53ee280a924>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.25-alpha.c2e748.0<br/>@codecademy/gamut-labs@19.5.3-alpha.c2e748.0<br/>@codecademy/gamut@34.6.3-alpha.c2e748.0<br/>@codecademy/styleguide@41.7.9-alpha.c2e748.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c8b99a696e3e00aec2a056--gamut-preview.netlify.app>https://60c8b99a696e3e00aec2a056--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c8b99a696e3e00aec2a056>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.4.26-alpha.21b353.0<br/>@codecademy/gamut-labs@19.6.1-alpha.21b353.0<br/>@codecademy/gamut@34.7.1-alpha.21b353.0<br/>@codecademy/styleguide@41.7.11-alpha.21b353.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c935e90b7a0e4896889b1b--gamut-preview.netlify.app>https://60c935e90b7a0e4896889b1b--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c935e90b7a0e4896889b1b>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.2.1-alpha.d77d11.0<br/>@codecademy/gamut-illustrations@0.17.6-alpha.d77d11.0<br/>@codecademy/gamut-kit@0.4.26-alpha.d77d11.0<br/>@codecademy/gamut-labs@19.6.1-alpha.d77d11.0<br/>@codecademy/gamut-patterns@0.4.1-alpha.d77d11.0<br/>@codecademy/gamut-styles@11.3.1-alpha.d77d11.0<br/>@codecademy/gamut-tests@2.4.9-alpha.d77d11.0<br/>@codecademy/gamut@34.7.1-alpha.d77d11.0<br/>@codecademy/konamimojisplosion@0.2.40-alpha.d77d11.0<br/>@codecademy/markdown-overrides@0.5.52-alpha.d77d11.0<br/>@codecademy/styleguide@41.7.11-alpha.d77d11.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c937a25af1884919208340--gamut-preview.netlify.app>https://60c937a25af1884919208340--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c937a25af1884919208340>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.2.1-alpha.15f475.0<br/>@codecademy/gamut-illustrations@0.17.6-alpha.15f475.0<br/>@codecademy/gamut-kit@0.4.26-alpha.15f475.0<br/>@codecademy/gamut-labs@19.6.1-alpha.15f475.0<br/>@codecademy/gamut-patterns@0.4.1-alpha.15f475.0<br/>@codecademy/gamut-styles@11.3.1-alpha.15f475.0<br/>@codecademy/gamut-tests@2.4.9-alpha.15f475.0<br/>@codecademy/gamut@34.7.1-alpha.15f475.0<br/>@codecademy/konamimojisplosion@0.2.40-alpha.15f475.0<br/>@codecademy/markdown-overrides@0.5.52-alpha.15f475.0<br/>@codecademy/styleguide@41.7.11-alpha.15f475.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c93a6de425484fce39c48c--gamut-preview.netlify.app>https://60c93a6de425484fce39c48c--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c93a6de425484fce39c48c>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.2.1-alpha.feb63c.0<br/>@codecademy/gamut-illustrations@0.17.6-alpha.feb63c.0<br/>@codecademy/gamut-kit@0.4.30-alpha.feb63c.0<br/>@codecademy/gamut-labs@19.8.1-alpha.feb63c.0<br/>@codecademy/gamut-patterns@0.4.1-alpha.feb63c.0<br/>@codecademy/gamut-styles@11.3.1-alpha.feb63c.0<br/>@codecademy/gamut-tests@2.4.9-alpha.feb63c.0<br/>@codecademy/gamut@34.7.2-alpha.feb63c.0<br/>@codecademy/konamimojisplosion@0.2.40-alpha.feb63c.0<br/>@codecademy/markdown-overrides@0.5.52-alpha.feb63c.0<br/>@codecademy/styleguide@41.9.1-alpha.feb63c.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c94e1c132bc10ea6085fb8--gamut-preview.netlify.app>https://60c94e1c132bc10ea6085fb8--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c94e1c132bc10ea6085fb8>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.2.1-alpha.8879eb.0<br/>@codecademy/gamut-illustrations@0.17.6-alpha.8879eb.0<br/>@codecademy/gamut-kit@0.4.30-alpha.8879eb.0<br/>@codecademy/gamut-labs@19.8.1-alpha.8879eb.0<br/>@codecademy/gamut-patterns@0.4.1-alpha.8879eb.0<br/>@codecademy/gamut-styles@11.3.1-alpha.8879eb.0<br/>@codecademy/gamut-tests@2.4.9-alpha.8879eb.0<br/>@codecademy/gamut@34.7.2-alpha.8879eb.0<br/>@codecademy/konamimojisplosion@0.2.40-alpha.8879eb.0<br/>@codecademy/markdown-overrides@0.5.52-alpha.8879eb.0<br/>@codecademy/styleguide@41.9.1-alpha.8879eb.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c9525f5ceee814f8d08eff--gamut-preview.netlify.app>https://60c9525f5ceee814f8d08eff--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c9525f5ceee814f8d08eff>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.2.1-alpha.bbd3fb.0<br/>@codecademy/gamut-illustrations@0.17.6-alpha.bbd3fb.0<br/>@codecademy/gamut-kit@0.4.30-alpha.bbd3fb.0<br/>@codecademy/gamut-labs@19.8.1-alpha.bbd3fb.0<br/>@codecademy/gamut-patterns@0.4.1-alpha.bbd3fb.0<br/>@codecademy/gamut-styles@11.3.1-alpha.bbd3fb.0<br/>@codecademy/gamut-tests@2.4.9-alpha.bbd3fb.0<br/>@codecademy/gamut@34.7.2-alpha.bbd3fb.0<br/>@codecademy/konamimojisplosion@0.2.40-alpha.bbd3fb.0<br/>@codecademy/markdown-overrides@0.5.52-alpha.bbd3fb.0<br/>@codecademy/styleguide@41.9.1-alpha.bbd3fb.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60c95acbb2f5541f54fa1754--gamut-preview.netlify.app>https://60c95acbb2f5541f54fa1754--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60c95acbb2f5541f54fa1754>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.2.1-alpha.4b8ba8.0<br/>@codecademy/gamut-illustrations@0.17.6-alpha.4b8ba8.0<br/>@codecademy/gamut-kit@0.4.30-alpha.4b8ba8.0<br/>@codecademy/gamut-labs@19.8.1-alpha.4b8ba8.0<br/>@codecademy/gamut-patterns@0.4.1-alpha.4b8ba8.0<br/>@codecademy/gamut-styles@11.3.1-alpha.4b8ba8.0<br/>@codecademy/gamut-tests@2.4.9-alpha.4b8ba8.0<br/>@codecademy/gamut@34.7.2-alpha.4b8ba8.0<br/>@codecademy/konamimojisplosion@0.2.40-alpha.4b8ba8.0<br/>@codecademy/markdown-overrides@0.5.52-alpha.4b8ba8.0<br/>@codecademy/styleguide@41.9.1-alpha.4b8ba8.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60ca0a08d2ad1ed93b798eea--gamut-preview.netlify.app>https://60ca0a08d2ad1ed93b798eea--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60ca0a08d2ad1ed93b798eea>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60ca15c77c9d3cf2493bc722--gamut-preview.netlify.app>https://60ca15c77c9d3cf2493bc722--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60ca15c77c9d3cf2493bc722>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.2.1-alpha.602279.0<br/>@codecademy/gamut-illustrations@0.17.6-alpha.602279.0<br/>@codecademy/gamut-kit@0.4.30-alpha.602279.0<br/>@codecademy/gamut-labs@19.8.1-alpha.602279.0<br/>@codecademy/gamut-patterns@0.4.1-alpha.602279.0<br/>@codecademy/gamut-styles@11.3.1-alpha.602279.0<br/>@codecademy/gamut-tests@2.4.9-alpha.602279.0<br/>@codecademy/gamut@34.7.2-alpha.602279.0<br/>@codecademy/konamimojisplosion@0.2.40-alpha.602279.0<br/>@codecademy/markdown-overrides@0.5.52-alpha.602279.0<br/>@codecademy/styleguide@41.9.1-alpha.602279.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60ca5b4a132bc13e7d085fcc--gamut-preview.netlify.app>https://60ca5b4a132bc13e7d085fcc--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60ca5b4a132bc13e7d085fcc>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-icons@5.2.2-alpha.50e568.0<br/>@codecademy/gamut-illustrations@0.17.7-alpha.50e568.0<br/>@codecademy/gamut-kit@0.4.31-alpha.50e568.0<br/>@codecademy/gamut-labs@19.8.2-alpha.50e568.0<br/>@codecademy/gamut-patterns@0.4.2-alpha.50e568.0<br/>@codecademy/gamut-styles@11.3.2-alpha.50e568.0<br/>@codecademy/gamut-tests@2.4.10-alpha.50e568.0<br/>@codecademy/gamut@34.8.1-alpha.50e568.0<br/>@codecademy/konamimojisplosion@0.2.41-alpha.50e568.0<br/>@codecademy/markdown-overrides@0.5.53-alpha.50e568.0<br/>@codecademy/styleguide@41.9.2-alpha.50e568.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://60ca6ac7bd894f00e6aacb01--gamut-preview.netlify.app>https://60ca6ac7bd894f00e6aacb01--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/60ca6ac7bd894f00e6aacb01>Deploy Logs</a></p>

**Pull request merged**